Index: subversion/libsvn_client/auth.c
===================================================================
--- subversion/libsvn_client/auth.c	(revision 5950)
+++ subversion/libsvn_client/auth.c	(working copy)
@@ -162,11 +162,9 @@
           apr_gid_t gid;
           apr_status_t status;
           
-          if ((status = apr_uid_current (&uid, &gid, pool)))
-            return svn_error_create (status, NULL, "Error getting UID");
-          if ((status = apr_uid_name_get (&un, uid, pool)))
-            return svn_error_create (status, NULL, "Error getting username");
-          SVN_ERR (svn_utf_cstring_to_utf8 (&def_username, un, NULL, pool));
+          if (!(status = apr_uid_current (&uid, &gid, pool)))
+            if (!(status = apr_uid_name_get (&un, uid, pool)))
+              SVN_ERR (svn_utf_cstring_to_utf8 (&def_username, un, NULL, pool));
         }
 
       def_password = apr_hash_get (parameters, 
